File: cmd/fem/solver/eigen.cpp
#include <Eigen/Sparse>
#ifdef __linux__
    #include <Eigen/PardisoSupport>
#else
    #include <Eigen/SparseCholesky>
#endif
#include <iostream>
#include "eigen.h"

using namespace Eigen;
using namespace std;

SparseMatrix<double> mat;
VectorXd vec;

void InitMatrix(int size, int max_non_zero)
{
    mat.resize(size, size);
    mat.setZero();
    vec.resize(size);
    vec.setZero();
    // Reserving memory
    mat.reserve(VectorXi::Constant(size, max_non_zero));
}

void SetMatrix(int row, int col, double value)
{
    mat.coeffRef(row, col) = value;
}

void AddMatrix(int row, int col, double value)
{
    mat.coeffRef(row, col) += value;
}

void SetVector(int i, double value)
{
    vec(i) = value;
}

void AddVector(int i, double value)
{
    vec(i) += value;
}

double GetMatrix(int row, int col)
{
    return mat.coeffRef(row, col);
}

double GetVector(int i)
{
    return vec(i);
}

void SetBoundaryCondition(int index, double value)
{
    for (Eigen::SparseMatrix<double>::InnerIterator i(mat, index); i; ++i)
    {
        if (i.row() not_eq i.col())
        {
            mat.coeffRef(i.row(), i.col()) = value;
            mat.coeffRef(i.col(), i.row()) = value;
        }
    }
    vec(index) = value * mat.coeffRef(index, index);
}

int SolveEigen(double *res)
{
#ifdef __linux__
    PardisoLLT<SparseMatrix<double>> solver;
#else
    SimplicialLLT<SparseMatrix<double>> solver;
#endif
    solver.compute(mat);
    if (solver.info() != Success)
        return 1;
    vec = solver.solve(vec);
    for (auto i = 0u; i < mat.rows(); i++)
        res[i] = vec[i];
    mat.resize(0, 0);
    vec.resize(0);
    return 0;
}

File: cmd/fem/solver/eigen.go
package solver

/*
//#cgo CXXFLAGS: -std=c++20 -I../../../../../eigen
#cgo CXXFLAGS: -std=c++20 -I../../../../../eigen -I/usr/include/mkl
#cgo LDFLAGS: -L/lib/x86_64-linux-gnu -lmkl_intel_lp64 -lmkl_sequential -lmkl_core
#include "eigen.h"
*/

import (
	"C"
	"fmt"
	"unsafe"
	"wfem/cmd/fem/mesh"
	"wfem/cmd/fem/progress"

	"gonum.org/v1/gonum/mat"
)

type EigenSolver struct {
	size int
}

func NewEigenSolver(mesh *mesh.Mesh) *EigenSolver {
	maxNonZero := 0
	for i := range mesh.MeshMap {
		if len(mesh.MeshMap[i]) > maxNonZero {
			maxNonZero = len(mesh.MeshMap[i])
		}
	}
	C.InitMatrix((C.int)(mesh.NumVertex()*mesh.Freedom()), (C.int)(2*maxNonZero*mesh.Freedom()))
	return &EigenSolver{size: mesh.NumVertex() * mesh.Freedom()}
}

func (_ *EigenSolver) SetMatrix(i, j int, value float64) {
	C.SetMatrix((C.int)(i), (C.int)(j), (C.double)(value))
}

func (_ *EigenSolver) AddMatrix(i, j int, value float64) {
	C.AddMatrix((C.int)(i), (C.int)(j), (C.double)(value))
}

func (_ *EigenSolver) SetVector(i int, value float64) {
	C.SetVector((C.int)(i), (C.double)(value))
}

func (_ *EigenSolver) AddVector(i int, value float64) {
	C.AddVector((C.int)(i), (C.double)(value))
}

func (_ *EigenSolver) GetMatrix(i, j int) float64 {
	return float64(C.GetMatrix((C.int)(i), (C.int)(j)))
}

func (es *EigenSolver) GetVector(i int) float64 {
	return float64(C.GetVector((C.int)(i)))
}

func (es *EigenSolver) SetBoundaryCondition(i int, value float64) {
	C.SetBoundaryCondition((C.int)(i), (C.double)(value))
}

func (es *EigenSolver) Solve() (*mat.VecDense, error) {
	var err error
	x := make([]float64, es.size)
	msg := progress.NewUnlimitedProgress("Solution of the system of equations")
	defer msg.StopProgress()
	result := C.SolveEigen((*C.double)(unsafe.Pointer(&x[0])))
	if result != 0 {
		err = fmt.Errorf("matrix is near singular")
	}
	return mat.NewVecDense(es.size, x), err
}
